readme.md
# Corridor

This is a p5.js tryout sketch, displaying a moving corridor made of frames
following random trajectories.

The goal is to link the corridor dynamically with audio input.

For now, the frames' brightness varies with the microphone input volume, and 
trajectory changes are triggered by volume peaks.

The microphone input with p5.js is limited to certain browsers. Access with
Chrome is limited to https.

 <img src="docs/example.png" width="50%" />


changelog.txt
Changelog for Corridor.


-------------------------------------------------------------------------------
20190721
    -added color and trajectory changes based on volume

-------------------------------------------------------------------------------
20190615
    -added sound module, with basic fft analysis

-------------------------------------------------------------------------------
20190608
    -added mic audio input

-------------------------------------------------------------------------------
20190519
    -added Trajectory sublasses

-------------------------------------------------------------------------------
20190510
    -added center as moving dot test script
    -added maximum distance for new frame center in corridor class
    -improved segment trajectory by allowing randomly outwards direction

-------------------------------------------------------------------------------
20190428
    -added segment trajectory
    -improved code structure
    -removed Square static variables

-------------------------------------------------------------------------------
20190427
    -added Square static variables
    -added Trajectory class


modules/test.js
/*
Processing.js tryouts
*/

// global variables
let canvas_width = 800;
let canvas_height = 600;


class Corridor{

    constructor(){
        this.velocity = 1.1;
        this.period= 3;
        this.center = [100, 200];
        this.trajectory = new Trajectory(`segment`, [100, 200], 0);
    }

    update(elapsed_time){
        /*
        Enlarge each frame, insert new small frames and remove large frames.
        */

        // insert new frames periodically
        if (elapsed_time % this.period == 0){
            this.center = this.trajectory.compute_new_point(elapsed_time)
        }

    }

    display(){
        ellipse(this.center[0], this.center[1], 10, 10)
    }

}


let elapsed_time = 0;
let my_corridor = new Corridor;


function setup() {
    /*
    Processing setup function.
    Called one time at start.
    */
    createCanvas(canvas_width, canvas_height);
    stroke(255);
    frameRate(30);
}


function draw() {
    /*
    Processing main loop function.
    */

    // initialisation
    translate(width/2, height/2);
    background(0);

    ellipse(0, 0, 10, 10)

    my_corridor.update(elapsed_time);
    my_corridor.display();

    elapsed_time += 1;

    if (elapsed_time % 100 == 0){
        my_corridor.trajectory = new Trajectory(
            `segment`,
            my_corridor.center,//my_corridor.frames[my_corridor.frames.length - 1].center,
            elapsed_time
        )
    } else if (elapsed_time % 100 == 50){
        my_corridor.trajectory = new Trajectory(
            `circle`,
            my_corridor.center,//my_corridor.frames[my_corridor.frames.length - 1].center,
            elapsed_time
        )
    }

}


tests/test_fft.js

function start_fft_sketch() {

    var fft_sketch = function(p5js) {
        /*
        Input:
            -p5js   p5js instance
        */

        let canvas_width = 800;
        let canvas_height = 600;
        let mic;

        let band = 3;

        p5js.setup = function() {

            mic = new p5.AudioIn();
            mic.start();
            my_analysis = new Analysis(band, 20);
            my_analysis.display_setup(canvas_width, canvas_height, mic, p5js);

        }

        p5js.draw = function() {

            p5js.background(0);

            bpm = my_analysis.analyze();

            my_analysis.display_spectrum(p5js);
            my_analysis.display_band_signal(p5js);
            my_analysis.display_threshold(p5js);

            p5js.text('click to play/pause', 4, 10);
            p5js.text('BPM: ' + bpm, 4, 20);
            p5js.text('band: ' + band, 4, 30);

        }

    }

    var myp5 = new p5(fft_sketch);

}


tests/test_maths.js




let test_list = [5.0, 2.0, 3.3, 5.6];

console.log(test_list)
console.log(normalize(2.4, test_list))


modules/sound.js
/*
Sound module for the Corridor project.
*/

class Analysis {
    /*
    A class for the analysis of an input sound.
    The input sound spectrum is obtained via FFT, and the BPM is measured on a
    specific band of this spectrum.
    */

    constructor(band, threshold) {
        this.band = band;
        this.threshold = threshold;
        this.size = 500;

        this.fft = new p5.FFT(0.8, 16);
        this.spectrum = [];
        this.amplitude_history = [0.0];
    }

    analyze() {
        /*
        Analyze input sound using FFT.
        Output:
            -bpm    float
        */

        this.spectrum = this.fft.analyze()

        // inserting new values in the amplitude history
        this.amplitude_history.push(this.spectrum[this.band])

        // removing old values if the maximum sized is reached
        if (this.amplitude_history.length > this.size) {
            this.amplitude_history.shift();
        }

        // BPM
        let count = 0;
        for (var i = 0; i < this.amplitude_history.length; i++){
            if (this.amplitude_history[i] > this.threshold) {
                count += 1;
            }
        }
        let bpm = count / this.amplitude_history.length;

        return bpm

    }

    display_setup(canvas_width, canvas_height, source, p5js){
        /*
        P5.js related setup function for this analysis display.
        Input:
            -canvas_width   int
            -canvas_height  int
            -source         p5.sound object
            -p5js           p5.js instance
        */
        this.canvas = p5js.createCanvas(canvas_width, canvas_height);
        this.fft.setInput(source)
    }

    display_spectrum(p5js) {
        /*
        Display spectrum of input sound.
        Input:
            -p5js   p5.js instance
        */
        p5js.noStroke();
        p5js.fill(0, 255, 0); // spectrum is green
        for (var i = 0; i < this.spectrum.length; i++){
            var x = p5js.map(i, 0, this.spectrum.length, 0, this.canvas.width);
            var h = -this.canvas.height + p5js.map(
                this.spectrum[i],
                0,
                255,
                this.canvas.height,
                0
            );
            p5js.rect(x, this.canvas.height, this.canvas.width / this.spectrum.length, h)
        }
    }

    display_band_signal(p5js) {
        /*
        Display amplitude signal of analyzed band.
        Input:
            -p5js   p5.js instance
        */
        p5js.noFill();
        p5js.beginShape();
        p5js.stroke(255, 0, 0);
        p5js.strokeWeight(1);
        for (var i = 0; i< this.amplitude_history.length; i++){
          var x = p5js.map(i, 0, this.amplitude_history.length, 0, this.canvas.width);
          var y = p5js.map( this.amplitude_history[i], 0, 255, 0, this.canvas.height);
          p5js.vertex(x, y);
        }
        p5js.endShape();
    }

    display_threshold(p5js) {
        /*
        Display threshold for this analysis to detect beats and count BPM.
        Input:
            -p5js   p5.js instance
        */
        p5js.noFill();
        p5js.stroke(0, 0, 255);
        p5js.strokeWeight(1);
        p5js.line(
            0,
            this.threshold*this.canvas.height/255,
            this.canvas.width,
            this.threshold*this.canvas.height/255,
        );
    }


}


modules/geometry.js
/*
Geometry module for the corridor project.
*/


/* Transformations ======================================================== */

function point_rotation(input, angle){
    /*
    Apply rotation around center to XY coordinates.
    Input:
        -input      X, Y coordinates
        -angle      radians
    Output:
        -output     X, Y coordinates
    */
    return [
        input[0] * Math.cos(angle) - input[1] * Math.sin(angle),
        input[1] * Math.cos(angle) + input[0] * Math.sin(angle)
    ]
}

function point_translation(input, vector){
    /*
    Apply translation.
    Input:
        -input      X, Y coordinates
        -vector     X, Y coordinates
    Output:
        -output     X, Y coordinates
    */
    return [
        input[0] + vector[0],
        input[1] + vector[1]
    ]
}


/* Trajectories =========================================================== */

function segment_through_center(
    start_point,
    elapsed_time,
    distance_max,
    direction_is_inward
){
    /*
    Describes segment with center at [0, 0] and one end at start_point.
    Input:
        -start_point            [X, Y]
        -elapsed_time
        -distance_max           float
            distance from center upper limit
        -direction_is_inward    bool
            defines if moving towards center or away from center at start
    Output:
        -new_point              [X, Y]
    */

    // rotate reference frame to have start_point on X-axis
    let angle = Math.atan2(start_point[1], start_point[0]);
    let start_point_proj = point_rotation(start_point, -angle);

    let new_point_proj_x;
    if (direction_is_inward == true){
        // time to travel from distance_max to center is set to proj_point X value.
        let time = elapsed_time % (4 * distance_max) + distance_max - start_point_proj[0];

        if (time < 2 * distance_max){
            // start_point - end_point portion
            new_point_proj_x = distance_max - time;
        } else {
            // end_point - start_point portion
            new_point_proj_x = - 3 * distance_max + time;
        }
    } else {
        // time to travel from distance_max to center is set to proj_point X value.
        if (elapsed_time < distance_max - start_point_proj[0]){
            new_point_proj_x = start_point_proj[0] + elapsed_time;
        } else {
            let time = (elapsed_time + start_point_proj[0] - distance_max) % (4 * distance_max);
            if (time < 2 * distance_max){
                // start_point - end_point portion
                new_point_proj_x = distance_max - time;
            } else {
                // end_point - start_point portion
                new_point_proj_x = - 3 * distance_max + time;
            }
        }
    }
    // return new point described in initial reference frame
    return point_rotation([new_point_proj_x, 0], angle)

}

function circle_equation(start_point, t){
    /*
    Describes circle.
    Input:
        -start_point
        -t
    Output:
        -new_point      [X, Y]
    */

    // rotate reference frame to have start_point on X-axis
    let angle = Math.atan2(start_point[1], start_point[0]);
    let projected_point = point_rotation(start_point, -angle);

    // compute point on circle in rotated reference frame
    let radius = projected_point[0]
    let new_projected_point = [
        radius * Math.cos(t),
        radius * Math.sin(t)
    ]

    // return new point described in initial reference frame
    return point_rotation(new_projected_point, angle)
}

class Trajectory{

    constructor(start_point, start_time, distance_max){
        /*
        Input:
            -start_point    [X, Y]
            -start_time     float
            -distance_max   float
                distance from center upper limit for this trajectory
        */

        this.start_point = start_point;
        this.start_time = start_time;
        this.distance_max = distance_max;

    }

}

class SegmentTraj extends Trajectory{

    constructor(start_point, start_time, distance_max){
        super(start_point, start_time, distance_max);
        this.direction_is_inward = Math.random() >= 0.5;
    }

    compute_new_point(elapsed_time){
        /*
        Input:
            -elapsed_time
        Output:
            -new_point      [X, Y]
        */

        return segment_through_center(
            this.start_point,
            elapsed_time - this.start_time,
            this.distance_max,
            this.direction_is_inward
        )

    }

}

class CircleTraj extends Trajectory{

    constructor(start_point, start_time, distance_max){
        super(start_point, start_time, distance_max);
    }

    compute_new_point(elapsed_time){
        /*
        Input:
            -elapsed_time
        Output:
            -new_point      [X, Y]
        */

        return circle_equation(
            this.start_point,
            (elapsed_time - this.start_time) * Math.PI/180
        )

    }

}


/* Shapes ================================================================= */

class Square{

    constructor(side, angle, center, size_max, color){
        /*
        Input:
            -side
            -angle      radians
            -center     [X, Y]
            -size_max
        */
        this.side = side;
        this.angle = angle;
        this.center = center;
        this.size_max = size_max;
        this.color = color;
    }

    display(p5js){
        /*
        Input:
            -p5js   p5.js instance
        */

        let corners = [
            [-this.side/2, -this.side/2],
            [-this.side/2, this.side/2],
            [this.side/2, this.side/2],
            [this.side/2, -this.side/2]
        ];

        for (let i = 0; i < corners.length; i++){
            corners[i] = point_rotation(corners[i], this.angle);
            corners[i] = point_translation(corners[i], this.center)
        }

        p5js.stroke(this.color);

        p5js.line(corners[0][0], corners[0][1], corners[1][0], corners[1][1]);
        p5js.line(corners[1][0], corners[1][1], corners[2][0], corners[2][1]);
        p5js.line(corners[2][0], corners[2][1], corners[3][0], corners[3][1]);
        p5js.line(corners[3][0], corners[3][1], corners[0][0], corners[0][1]);
    }

}


sketch.js
/*
Processing.js tryouts
*/

// global variables
let canvas_width = 800;
let canvas_height = 600;
let mic;
let frame_rate = 30; //frames per second
let volume_history = [0.0];

class Corridor{

    constructor(){
        this.velocity = 1.1; // ratio at which frames grow bigger
        this.rotational_velocity = 0.0;
        this.period = 3; // rate at which new frames are inserted
        this.distance_max = 200; // distance from center upper limit
        this.frames = [];
        this.trajectory = new SegmentTraj(
            [100, 200], // start point
            0, // start time
            this.distance_max // max distance
        );
    }

    update(elapsed_time, new_frames_color){
        /*
        Enlarge each frame, insert new small frames and remove large frames.
        Input:
            -elapsed_time       float
            -new_frames_color   float
        */

        // update each frame
        for (let i = 0; i < this.frames.length; i++){

            // make frames larger
            this.frames[i].side *= this.velocity;

        }

        // insert new frames periodically
        if (elapsed_time % this.period == 0){
            this.frames.push(new Square(
                5, // side
                elapsed_time * this.rotational_velocity, // angle
                this.trajectory.compute_new_point(elapsed_time), // center
                Math.sqrt(2) * Math.max(canvas_width, canvas_height), // size_max
                new_frames_color
            ));
        }

        // get rid of first frame if larger than max size
        if (this.frames[0].side > this.frames[0].size_max){
            this.frames.shift();
        }

    }

    display(p5js){
        /*
        Input:
            -p5js     p5js instance
        */
        for (let i = 0; i < this.frames.length; i++){
            this.frames[i].display(p5js);
        }
    }

}


function start_corridor_sketch() {

    var corridor_sketch = function(p5js) {
        /*
        Input:
            -p5js   p5js instance
        */

        let elapsed_time_frames = 0;
        let my_corridor = new Corridor;

        p5js.setup = function() {
            /*
            Processing setup function.
            Called one time at start.
            */
            var canvas = p5js.createCanvas(canvas_width, canvas_height);
            canvas.parent("canvas_div");
            p5js.stroke(255);
            p5js.frameRate(frame_rate);
            mic = new p5.AudioIn();
            mic.start();
        };

        p5js.draw = function() {
            /*
            Processing main loop function.
            */

            // initialisation
            p5js.translate(p5js.width/2, p5js.height/2);
            p5js.background(0);

            // adjust new_frames_color to mic volume
            let current_volume = mic.getLevel();
            volume_history.push(current_volume);
            if (volume_history.length > 200) {
                volume_history.shift()
            }
            current_volume_normalized = normalize(current_volume, volume_history);
            let new_frames_color = 255*current_volume_normalized;
            my_corridor.update(elapsed_time_frames, new_frames_color);
            my_corridor.display(p5js);

            elapsed_time_frames += 1;

            // change trajectory type when high volume is reached
            if (current_volume_normalized == 1){

                if (my_corridor.trajectory instanceof SegmentTraj){
                    my_corridor.trajectory = new CircleTraj(
                        my_corridor.frames[my_corridor.frames.length - 1].center,
                        elapsed_time_frames,
                        my_corridor.distance_max
                    )
                } else if (my_corridor.trajectory instanceof CircleTraj){
                    my_corridor.trajectory = new SegmentTraj(
                        my_corridor.frames[my_corridor.frames.length - 1].center,
                        elapsed_time_frames,
                        my_corridor.distance_max
                    )
                };

            };

        };

    };

    var myp5 = new p5(corridor_sketch);

};
